core/middlewares/with-auth.ts

```diff
@@ -1,17 +1,68 @@
-import { auth, signIn } from '~/auth';
+import { NextResponse } from 'next/server';
+
+import { auth, signIn, signOut } from '~/auth';
+import { client } from '~/client';
+import { graphql } from '~/client/graphql';
 
 import { type MiddlewareFactory } from './compose-middlewares';
 
+// Path matcher for any routes that require authentication
+const protectedPathRegExp = /^\/(account).*$/;
+
+async function checkTokenValidity(customerAccessToken: string, email?: string): Promise<boolean> {
+  const result = await client.fetch({
+    document: graphql(`
+      query ValidateToken {
+        customer {
+          email
+        }
+      }
+    `),
+    customerAccessToken,
+    fetchOptions: { cache: 'no-store' },
+  });
+
+  return result.data.customer?.email === email;
+}
+
+function redirectToLogin(url: string) {
+  return NextResponse.redirect(new URL('/login', url), { status: 302 });
+}
+
 export const withAuth: MiddlewareFactory = (next) => {
   return async (request, event) => {
     // @ts-expect-error: The `auth` function doesn't have the correct type to support it as a MiddlewareFactory.
     const authWithCallback = auth(async (req) => {
+      const isProtectedRoute = protectedPathRegExp.test(req.nextUrl.pathname);
```

Comment:
Instead of using regex, could we try using [`URLPattern`](https://developer.mozilla.org/en-US/docs/Web/API/URLPattern/URLPattern) instead? If that is a bit cumbersome here (since you'd need to pass in a hostname when constructing the object), then maybe we can just use hardcoded paths in an array.